plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +302,243 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
 	}
 }
 
+static void
+os_upgrade_cancelled_cb (GCancellable *cancellable,
+			 GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	eos_updater_call_cancel (priv->updater_proxy, NULL, NULL, NULL);
+}
+
+static void
+setup_os_upgrade_cancellable (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	GCancellable *cancellable = gs_app_get_cancellable (get_os_upgrade (plugin));
+
+	if (priv->os_upgrade_cancellable == cancellable)
+		return;
+
+	g_clear_object (&priv->os_upgrade_cancellable);
+	priv->os_upgrade_cancellable = g_object_ref (cancellable);
+	g_cancellable_connect (priv->os_upgrade_cancellable,
+			       G_CALLBACK (os_upgrade_cancelled_cb),
+			       plugin, NULL);
+}
+
+static void
+os_upgrade_set_download_by_user (GsApp *app, gboolean value)
+{
+	g_autoptr(GVariant) var = g_variant_new_boolean (value);
+	/* we need to assign it to NULL in order to be able to override it
+	 * (safeguard mechanism in GsApp...) */
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", NULL);
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", var);
+}
+
+static gboolean
+os_upgrade_get_download_by_user (GsApp *app)
+{
+	GVariant *value = gs_app_get_metadata_variant (app, "eos::DownloadByUser");
+	if (value == NULL)
+		return FALSE;
+	return g_variant_get_boolean (value);
+}
+
+static void
+os_upgrade_force_fetch (EosUpdater *updater)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
```

Comment:
What does the `force` flag do here, and why do we need it?

Reply:
The normal `Fetch()` method will check whether `libmogwai` allows it to perform the download or not. Since we only call fetch ourselves when the user clicks `Download`, I think we should force the download (hence the flag). Any verification (like a dialog confirming that the user wants to proceed even on a metered connection) should be done outside of this plugin.